tests/test_pipeline.py

```diff
@@ -0,0 +1,97 @@
+import os
+import pytest
+import pandas as pd
+import numpy as np
+from skimage.io import imsave
+import luigi
+from cyto_ml.pipeline.pipeline_decollage import ReadMetadata, DecollageImages, UploadDecollagedImagesToS3
+
+
+@pytest.fixture
+def temp_dir(tmp_path):
+    """Creates a temporary directory using pytest's tmp_path fixture."""
+    return tmp_path
+
+
+def test_read_metadata(temp_dir):
+    # Create a mock .lst file for testing
+    lst_file_content = "001\nnum-fields|value\n"
+    # The reader has to assume 53 "header" lines
+    for i in range(1, 54):
+        lst_file_content += f"field{i}|val{i}\n"
+    # Also assumes at least one row of data! Add two of them
+    for i in [0,1]:
+        lst_file_content += "|".join([str(i) for i in range(1,54)]) + "\n"
+
+    lst_file_path = os.path.join(temp_dir, 'test.lst')
+    with open(lst_file_path, 'w') as f:
+        f.write(lst_file_content)
+
+
+    # Run the ReadMetadata task
+    task = ReadMetadata(directory=str(temp_dir))
+    luigi.build([task], local_scheduler=True)
+
+    # Check if metadata.csv was created
+    output_file = task.output().path
+    assert os.path.exists(output_file), "Metadata CSV file should be created."
+    df = pd.read_csv(output_file)
+    print(df)
+    assert len(df) == 2, "The metadata CSV should have two fields."
+
+
+def test_decollage_images(temp_dir):
+    # Create mock metadata
+    metadata = pd.DataFrame({
+        "collage_file": ["test_collage.tif"],
+        "image_x": [0],
+        "image_y": [0],
+        "image_h": [100],
+        "image_w": [100]
+    })
+    metadata.to_csv(os.path.join(temp_dir, "metadata.csv"), index=False)
+
+    # Create a mock TIFF image
+    img_path = os.path.join(temp_dir, "test_collage.tif")
+    img = np.zeros((200, 200), dtype=np.uint8)
+    imsave(img_path, img)
+
+    # Run the DecollageImages task
+    task = DecollageImages(directory=str(temp_dir), output_directory=str(temp_dir), experiment_name="test_experiment")
+    luigi.build([task], local_scheduler=True)
+
+    # Check if the output image was created
+    output_image = os.path.join(temp_dir, "test_experiment_0.tif")
+    assert os.path.exists(output_image), "Decollaged image should be created."
+
+class MockTask(luigi.Task):
+    directory = luigi.Parameter()
+    check_unfulfilled_deps = False
+    def output(self) -> luigi.Target:
+        # "The output() method returns one or more Target objects.""
+        return luigi.LocalTarget(f'{self.directory}/out.txt') 
+    
+def test_upload_to_api(temp_dir, mocker):
+    # Write a tmp file to serve as our upstream task's output
+    with open(os.path.join(temp_dir, 'out.txt'), 'w') as out:
+        out.write("blah")
+    # The task `requires` DecollageImages, but that requires other tasks, which run first
+    # Rather than mock its output, or the whole chain, require a mock task that replaces it 
+    mock_output = mocker.patch(f'cyto_ml.pipeline.pipeline_decollage.UploadDecollagedImagesToS3.requires')
```

Comment:
This one was more interesting! As it [stood previously](https://github.com/NERC-CEH/plankton_pipeline_luigi/blob/6136cafd6d25d767623056ad6d37f3915c8ac84c/test/test_pipeline.py#L64) the code had mocked output for the pipeline task that this one `requires`

That output was a `str` where a subclass of `luigi.Target` was required (and the task runner triggered a failure because of that). Tried replacing the output with a `MockTarget` but then all the sub-tasks that `DecollageImages` has in its `requires` function started to run before it ever got to returning its mocked `output`

So I did this instead, replaced the whole `requires` chain with a single `MockTask` that returns an arbitrary file output, which we create first.

This was a good path into understanding Luigi! These complex dependencies with different data stages are hard to unit test like this, I'm impressed by the subtlety with which you did it. In your place I'd probably have written a big baggy "integration"-style test running the whole task graph and using `moto-server`, and that would have been much less efficient :D